doc/mm.rst

```diff
@@ -0,0 +1,101 @@
+=======================
+Nim's Memory Management
+=======================
+
+.. default-role:: code
+.. include:: rstcommon.rst
+
+:Author: Andreas Rumpf
+:Version: |nimversion|
+
+..
+
+
+  "The road to hell is paved with good intentions."
+
+
+Introduction
+============
+
+A memory-management algorithm optimal for every use-case cannot exist.
+Nim provides multiple paradigms for needs ranging from large multi-threaded
+applications, to games, hard-realtime systems and small microcontrollers.
+
+This document describes how the management strategies work;
+How to tune the garbage collectors for your needs, like (soft) `realtime systems`:idx:,
+and how the memory management strategies other than garbage collectors work.
+
+
+Multi-paradigm Memory Management Strategies
+===========================================
+
+.. default-role:: option
+
+To choose the memory management strategy use the `--mm:` switch.
+
+**The recommended switch is `--mm:orc`.**
```

Comment:
this recommendation seems a bit over the top, given the compiler itself doesn't use it

Reply:
also, orc/arc bugs are still quite common, and quite severe (https://github.com/nim-lang/Nim/issues?q=is%3Aissue+is%3Aopen+arc+label%3AARC%2FORC) - is the "default" gc worse / less stable?

Reply:
IME the default GC is less stable and its unresolved bugs are more severe than Orc's. But of course, you can disagree.

Reply:
well, it's not the default, and that deserves an explanation - one such explanation is that code likely needs to be updated to even work with orc/arc , if it does anything beyond hello world - in that sense, the framing here would be that orc/arc is expected to become the default for 2.0 (seeing the number of "make it work with orc" commits in various libraries that have already been done) - it's also very different when you're developing a small script / utility vs a library that others will use - in the latter world, you don't know what flags _users_ of your library will pass so you need to be aware of both models.

Given the default is refc, "normal" use of the compiler will be refc for the foreseeable future, and this is something to keep in mind (specially for people using 1.6): the "I downloaded and ran nim for the first time" experience is _still_ refc, and that's the impression people will walk away with, _unless_ they hang out on the nim dev forum and read articles about memory management models.

One other relevant point is that developing well-written code for ORC looks different than developing well-written code for refc - the former has different finalization semantics (ie _sometimes_ the finalizers are run more deterministically but still not quite, so files will close and memory will be released at a different time), moves and so on.